### PR TITLE
Radio contact with followers, NPC overmap movement.

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Radio contact with followers, NPC overmap movement"```

#### Purpose of change
If you and your follower both have a two-way radio, you can select them in your follower menu, and talk to them.
All I've enabled in that menu so far is "go to location" the location being either a) player location or b) location of one of your camps. and " stay where you are " so they stop moving.
The NPC will then path to that location on the overmap and inform you when they are there.
( this functionality has also been added to the standard talk menu )
There is a debug menu option to show NPC path as an overlay.
NPC will use the A* to find best overmap route ( they prefer roads )

This is the first step towards https://github.com/CleverRaven/Cataclysm-DDA/projects/7 and https://github.com/CleverRaven/Cataclysm-DDA/issues/28393

I have only enabled NPC followers to be ordered to move, but theoretically all dynamic NPCs could be allowed to travel on the overmap now.

This is a rough first-pass. there are some caveats and limitations.
1. Radios don't use batteries when used to give orders.
2. Radios don't have range currently - rising a z-level, being indoors, having a radio tower at your camp, all these things could potentially change radio range.
3. NPCs kinda just pop-into existence if you are there in the destination location to see them arrive, this is beyond my ability to fix, due to the the NPC spawning code using submaps.
4. There is no risk to the NPC whilst travelling, some kind of risk calculation would be good, aswell as a speed calculation to determine the terrain and the NPCs skills.
5. the A* algorithm is very basic in terms of OMT terrain costs, this could be refined. ( roads are preferred to fields which are preferred to rivers and forests and swamps, basically. )
6. z-level weirdness may occur, althoug in testing they seemed to cope with going up or down z-levels fine.

#### Describe the solution
Added code to check for a radio with follower and player, in faction menu.
made a new talk menu for radio contact.
made a new mission type, making use of existing pathfinding omt goal that NPCs have.
added code to do_turn() that checks for follower NPCs with travelling missions.
do some pathfinding, and spawn them in appropiate submaps every x minutes.
ensured that overmapbuffer keeps track if they switch overmap.
added some more robust follower validation and serialization for player followers ( pathfinding may take them too far away for the standard validation to find )

#### Describe alternatives you've considered
Had lots of ideas for other things, but its already getting too big, and wanted the first pass out the door for feedback.

#### Additional context
Things this may enable further work on :
Radio towers in camps.
vehicle mounted radio transmitters for further range.
remote administration of a basecamp
radio contact with non-follower faction NPCs ( old guard etc )

image

image

( NPC takes path through forest trail instead of through forest )

### CODE REVIEW

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/i-was-wonder-if-there-will-have-walkie-talkie-on-the-game-and-how-about-riot-shield/16434/13

**Comment by davidpwbrown** [comment]
jenkins rebuild

File: data/json/npcs/TALK_COMMON_ALLY.json [hunk 1]
```diff
@@ -377,6 +378,28 @@
       { "text": "Nah, I'm just kidding.", "topic": "TALK_NONE" }
     ]
   },
+  {
+    "id": "TALK_RADIO",
+    "type": "talk_topic",
+    "dynamic_line": "*pshhhttt* I'm reading you boss, over.",
+    "responses": [
+      { "text": "Please go to this location...", "topic": "TALK_GOTO_LOCATION_RADIO", "effect": "goto_location" },
+      { "text": "Stay at your current position", "topic": "TALK_DONE", "effect": "assign_guard" },
```
Comment (Night-Pryanik) [inline data/json/npcs/TALK_COMMON_ALLY.json]:
```suggestion
      { "text": "Stay at your current position.", "topic": "TALK_DONE", "effect": "assign_guard" },
```

File: data/json/npcs/TALK_COMMON_ALLY.json [hunk 2]
```diff
@@ -377,6 +378,28 @@
...
+      { "text": "Never mind.", "topic": "TALK_DONE" }
+    ]
+  },
+  {
+    "id": "TALK_GOTO_LOCATION_RADIO",
+    "type": "talk_topic",
+    "dynamic_line": "*pssshht* ten-four, I'll make my way there, over",
```
Comment (Night-Pryanik) [inline data/json/npcs/TALK_COMMON_ALLY.json]:
```suggestion
    "dynamic_line": "*pssshht* ten-four, I'll make my way there, over.",
```

File: data/json/npcs/TALK_COMMON_ALLY.json [hunk 3]
```diff
@@ -377,6 +378,28 @@
...
+    "responses": [ { "text": "Affirmative.", "topic": "TALK_DONE" } ]
+  },
+  {
+    "id": "TALK_GOTO_LOCATION",
+    "type": "talk_topic",
+    "dynamic_line": "Sure thing, I'll make my way there",
```
Comment (Night-Pryanik) [inline data/json/npcs/TALK_COMMON_ALLY.json]:
```suggestion
    "dynamic_line": "Sure thing, I'll make my way there.",
```

File: data/json/tool_qualities.json [hunk 4]
```diff
@@ -116,6 +116,11 @@
     "id": "SCREW_FINE",
     "name": "fine screw driving"
   },
+  {
+    "type": "tool_quality",
+    "id": "TWO_WAY_RADIO",
+    "name": "two way radio"
```
Comment (Night-Pryanik) [inline data/json/tool_qualities.json]:
```suggestion
    "name": "two-way radio"
```

File: src/faction.cpp [hunk 5]
```diff
@@ -1243,12 +1244,36 @@ void new_faction_manager::display() const
...
+                                if( temp_camp ) {
+                                    dest_camp = *temp_camp;
+                                    std::string camp_string = dest_camp->camp_name();
+                                    dest_string = _( "travelling to : " ) + camp_string;
+                                } else {
+                                    std::ostringstream oss;
+                                    oss << _( "travelling to : (" ) << dest->x << "," << dest->y << ")";
+                                    dest_string = oss.str();
+                                }
+                                mission_string = _( "Current Mision : " ) + dest_string;
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                mission_string = _( "Current Mission : " ) + dest_string;
```

File: src/faction.cpp [hunk 6]
```diff
@@ -1243,12 +1244,36 @@ void new_faction_manager::display() const
...
+                            } else {
+                                mission_string = _( "Current Mision : " ) + get_mission_action_string( c_mission.mission_id );
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                mission_string = _( "Current Mission : " ) + get_mission_action_string( c_mission.mission_id );
```

File: src/faction.cpp [hunk 7]
```diff
@@ -1259,23 +1284,45 @@ void new_faction_manager::display() const
...
+                                    radio_interactable = true;
+                                    can_see = "Within radio range";
+                                    see_color = c_light_green;
+                                } else {
+                                    can_see = "Not within radio range";
+                                    see_color = c_light_red;
+                                }
+                            } else if( guy->has_item_with_flag( "TWO_WAY_RADIO", true ) &&
+                                       !g->u.has_item_with_flag( "TWO_WAY_RADIO", true ) ) {
+                                can_see = "you do not have a radio";
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                can_see = "You do not have a radio";
```

File: src/faction.cpp [hunk 8]
```diff
@@ -1259,23 +1284,45 @@ void new_faction_manager::display() const
...
+                                see_color = c_light_red;
+                            } else if( !guy->has_item_with_flag( "TWO_WAY_RADIO", true ) &&
+                                       g->u.has_item_with_flag( "TWO_WAY_RADIO", true ) ) {
+                                can_see = "follower does not have a radio";
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                can_see = "Follower does not have a radio";
```

File: src/faction.cpp [hunk 9]
```diff
@@ -1259,23 +1284,45 @@ void new_faction_manager::display() const
...
+                                see_color = c_light_red;
+                            } else {
+                                can_see = "both you and follower need a radio";
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                can_see = "Both you and follower need a radio";
```

File: src/faction_camp.cpp [hunk 10]
```diff
@@ -100,6 +100,12 @@ std::map<std::string, bcp_miss_data> basecamp_missions_info = {{
                 " (Finish) Crafting", _( " (Finish) Crafting" )
             }
         },
+        {
+            "travelling", {
+                "Travelling", _( "Travelling" ), _( "Busy travelling!\n" ),
+                " Recall ally from travelling", _( "Recall ally from travelling" )
```
Comment (Night-Pryanik) [inline src/faction_camp.cpp]:
```suggestion
                "Recall ally from traveling", _( "Recall ally from traveling" )
```

File: src/faction_camp.cpp [hunk 11]
```diff
@@ -100,6 +100,12 @@ std::map<std::string, bcp_miss_data> basecamp_missions_info = {{
                 " (Finish) Crafting", _( " (Finish) Crafting" )
             }
         },
+        {
+            "travelling", {
+                "Travelling", _( "Travelling" ), _( "Busy travelling!\n" ),
```
Comment (Night-Pryanik) [inline src/faction_camp.cpp]:
```suggestion
                "Traveling", _( "Traveling" ), _( "Busy traveling!\n" ),
```

File: src/faction.cpp [hunk 12]
```diff
@@ -1243,12 +1244,36 @@ void new_faction_manager::display() const
...
                             npc_companion_mission c_mission = guy->get_companion_mission();
-                            mission_string = _( "Current Mision : " ) + get_mission_action_string( c_mission.mission_id );
+                            cata::optional<tripoint> dest = guy->get_mission_destination();
+                            if( dest ) {
+                                basecamp *dest_camp;
+                                cata::optional<basecamp *> temp_camp = overmap_buffer.find_camp( dest->x, dest->y );
+                                if( temp_camp ) {
+                                    dest_camp = *temp_camp;
+                                    std::string camp_string = dest_camp->camp_name();
+                                    dest_string = _( "travelling to : " ) + camp_string;
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                    dest_string = _( "traveling to : " ) + camp_string;
```

File: src/faction.cpp [hunk 13]
```diff
@@ -1243,12 +1244,36 @@ void new_faction_manager::display() const
...
+                                } else {
+                                    std::ostringstream oss;
+                                    oss << _( "travelling to : (" ) << dest->x << "," << dest->y << ")";
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                                    oss << _( "traveling to : (" ) << dest->x << "," << dest->y << ")";
```

File: src/faction.cpp [hunk 14]
```diff
@@ -1259,23 +1284,45 @@ void new_faction_manager::display() const
...
                         std::string can_see;
                         nc_color see_color;
-                        const std::vector<npc *> interactable_followers = g->get_npcs_if( [&]( const npc & guy ) {
-                            return ( ( guy.is_friend() && guy.is_following() ) || guy.mission == NPC_MISSION_GUARD_ALLY ) &&
-                                   g->u.posz() == guy.posz() &&
-                                   g->u.sees( guy.pos() ) && rl_dist( g->u.pos(), guy.pos() ) <= SEEX * 2;
-                        } );
-                        if( std::find( interactable_followers.begin(), interactable_followers.end(),
-                                       guy ) != interactable_followers.end() ) {
+                        // TODO NPCS on mission contactable same as travelling
```
Comment (Night-Pryanik) [inline src/faction.cpp]:
```suggestion
                        // TODO NPCS on mission contactable same as traveling
```

File: src/game.cpp [hunk 15]
```diff
@@ -3297,6 +3313,13 @@ void game::debug()
             break;
         }
         case 37:
+            if( debug_pathfinding ) {
+                debug_pathfinding = false;
+            } else {
+                debug_pathfinding = true;
```
Comment (Night-Pryanik) [inline src/game.cpp]: The whole thing could be rewritten as `debug_pathfinding = !debug_pathfinding`.

File: src/game.cpp [hunk 16]
```diff
@@ -4615,6 +4638,42 @@ void game::monmove()
     cleanup_dead();
 }
 
+void game::overmap_npc_move()
+{
+    std::vector<npc *> travelling_npcs;
+    // for now just processing NPC followers on travelling missions
```
Comment (Night-Pryanik) [inline src/game.cpp]:
```suggestion
    // for now just processing NPC followers on traveling missions
```

File: src/npcmove.cpp [hunk 17]
```diff
@@ -3062,6 +3062,23 @@ bool npc::has_omt_destination() const
...
+        guard_pos = global_square_location();
+        if( has_companion_mission() ) {
+            reset_companion_mission();
+        }
+        omt_path.clear();
+        goal = no_goal_point;
+        if( rl_dist( g->u.pos(), pos() ) > SEEX * 2 || !g->u.sees( pos() ) ) {
+            if( g->u.has_item_with_flag( "TWO_WAY_RADIO", true ) &&
+                has_item_with_flag( "TWO_WAY_RADIO", true ) ) {
+                add_msg( m_info, _( "From your Two-way radio you hear %s reporting in, 'I've arrived, boss!'" ),
```
Comment (Night-Pryanik) [inline src/npcmove.cpp]:
```suggestion
                add_msg( m_info, _( "From your two-way radio you hear %s reporting in, 'I've arrived, boss!'" ),
```

File: src/npctalk_funcs.cpp [hunk 18]
```diff
@@ -132,6 +132,52 @@ void talk_function::start_trade( npc &p )
     trade( p, 0, _( "Trade" ) );
 }
 
+void talk_function::goto_location( npc &p )
+{
+    int i = 0;
+    uilist selection_menu;
+    selection_menu.text = string_format( _( "Select a destination." ) );
```
Comment (Night-Pryanik) [inline src/npctalk_funcs.cpp]:
```suggestion
    selection_menu.text = string_format( _( "Select a destination" ) );
```

File: src/game.cpp [hunk 19]
```diff
@@ -3297,6 +3313,8 @@ void game::debug()
             break;
         }
         case 37:
+            debug_pathfinding = !debug_pathfinding;
```
Comment (Night-Pryanik) [inline src/game.cpp]: `break` has gone missing.

File: data/json/tool_qualities.json [hunk 20]
```diff
@@ -116,6 +116,11 @@
     "id": "SCREW_FINE",
     "name": "fine screw driving"
   },
+  {
+    "type": "tool_quality",
+    "id": "TWO_WAY_RADIO",
+    "name": "two-way radio"
```
Comment (Night-Pryanik) [inline data/json/tool_qualities.json]: I may be missing something, but is this used anywhere?

Comment (davidpwbrown) [reply data/json/tool_qualities.json]: I added the [two-way radio] tool flag. currently the only thing that uses it, is the two-way radio item, but in the future different items could use that flag too.
I thought if you add a tool flag you have to also add it to tool_qualities.json to describe it.

Comment (Night-Pryanik) [reply data/json/tool_qualities.json]: There's no need to add a tool quality for two-way radios - we don't have qualities for thermometers, for example, or watches. Describe it in `flags.json`.

**Comment by adamkad1** [comment]
OBEY ME!

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/yelling-commands-and-using-military-hand-signals/19487/7

File: src/game.cpp [hunk 21]
```diff
@@ -1843,12 +1847,19 @@ void game::add_npc_follower( const int &id )
...
+} ) ) {
+        u.follower_ids.push_back( id );
+    }
 }
 
 void game::remove_npc_follower( const int &id )
 {
     follower_ids.erase( std::remove( follower_ids.begin(), follower_ids.end(), id ),
                         follower_ids.end() );
+    follower_ids.erase( std::remove( u.follower_ids.begin(), u.follower_ids.end(), id ),
```
Comment (ZhilkinSerg) [inline src/game.cpp]: Possible cause of https://github.com/CleverRaven/Cataclysm-DDA/issues/29278.

I think it should be:

```suggestion
    u.follower_ids.erase( std::remove( u.follower_ids.begin(), u.follower_ids.end(), id ),
```

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/latest-experimental-features/5582/1106

**Pull request merged**